File: 2023/d14.cpp
#include <iostream>
#include <vector>
#include <string>
#include <set>

using namespace std;

vector<string> p;
int H, W;

void roll(int x, int y, int dx, int dy)
{
    int jx = x, jy = y;
    while (0 <= y && y < H && 0 <= x && x < W)
    {
        if (p[y][x] == '.')
        {
            if (dx + dy > 0)
            {
                jx = max(jx, x + dx);
                jy = max(jy, y + dy);
            }
            else
            {
                jx = min(jx, x + dx);
                jy = min(jy, y + dy);
            }
            while (0 <= jy && jy < H && 0 <= jx && jx < W)
            {
                if (p[jy][jx] == '#')
                {
                    x = jx;
                    y = jy;
                    break;
                }
                if (p[jy][jx] == 'O')
                {
                    swap(p[y][x], p[jy][jx]);
                    break;
                }
                jx += dx;
                jy += dy;
            }
        }
        x += dx;
        y += dy;
    }
}

void cycle()
{
    for (int x = 0; x < W; x++)
        roll(x, 0, 0, 1);
    for (int y = 0; y < H; y++)
        roll(0, y, 1, 0);
    for (int x = 0; x < W; x++)
        roll(x, H - 1, 0, -1);
    for (int y = 0; y < H; y++)
        roll(W - 1, y, -1, 0);
}

int calc()
{
    int res = 0;
    for (int y = 0; y < H; y++)
        for (int x = 0; x < W; x++)
            if (p[y][x] == 'O')
                res += H - y;
    return res;
}

int main()
{

    string row;
    while (getline(cin, row))
        p.push_back(row);
    H = p.size();
    W = p[0].size();

    for (int x = 0; x < W; x++)
        roll(x, 0, 0, 1);

    cout << "Part 1: " << calc() << endl;

    int i;
    set<string> seen;
    vector<string> prev;
    vector<int> scores;

    while (true)
    {
        row = "";
        for (i = 0; i < H; i++)
            row += p[i];
        if (seen.find(row) != seen.end())
            break;
        prev.push_back(row);
        seen.insert(row);
        scores.push_back(calc());
        cycle();
    }

    int offset;
    for (offset = 0; offset < prev.size(); offset++)
        if (prev[offset] == row)
            break;
    int repeat = scores.size() - offset;

    cout << "Part 2: " << scores[((1000000000 - offset) % repeat) + offset] << endl;

    return 0;
}

File: 2023/d14.py
from aoc import *


def parse(r):
    return [0 if c == "O" else (1 if c == "#" else 2) for c in r]


data = read(14, ["\n", parse])

H, W = len(data), len(data[0])


def roll(x, y, dx, dy, data=data, H=H, W=W):
    jx, jy = x, y
    mf = max if dx + dy > 0 else min
    while 0 <= y and y < H and 0 <= x and x < W:
        if data[y][x] == 2:
            jx = mf(jx, x + dx)
            jy = mf(jy, y + dy)
            while 0 <= jy and jy < H and 0 <= jx and jx < W:
                t = data[jy][jx]
                if t == 1:
                    x, y = jx, jy
                    break
                if t == 0:
                    data[jy][jx] = data[y][x]
                    data[y][x] = t
                    break
                jx += dx
                jy += dy
        x += dx
        y += dy


def north():
    for x in range(W):
        roll(x, 0, 0, 1)


def cycle():
    north()
    for y in range(H):
        roll(0, y, 1, 0)
    for x in range(W):
        roll(x, H - 1, 0, -1)
    for y in range(H):
        roll(W - 1, y, -1, 0)


def calc():
    return sum(data[y].count(0) * (H - y) for y in range(H))


north()
print("Part 1:", calc())


history = set()
h = []
ch = []
while True:
    s = str(data)
    if s in history:
        break
    history.add(s)
    h.append(s)
    ch.append(calc())
    cycle()

offset = h.index(s)
repeat = len(history) - offset
print("Part 2:", ch[((1000000000 - offset) % repeat) + offset])
